zookeeper-server/src/main/java/org/apache/zookeeper/server/Request.java

```diff
@@ -164,6 +169,26 @@ public boolean isThrottlable() {
                 && this.type != OpCode.createSession;
     }
 
+    private transient byte[] serializeData;
+
+    @SuppressFBWarnings(value = "EI_EXPOSE_REP")
+    public byte[] getSerializeData() {
+        if (this.hdr == null) {
+            return null;
+        }
+
+        if (this.serializeData == null) {
+            try {
+                this.serializeData = Util.marshallTxnEntry(this.hdr, this.txn, this.txnDigest);
+            } catch (IOException e) {
+                LOG.error("This really should be impossible.", e);
```

Comment:
Still I'm thinking of if we can assert this case won't happen. But it's an inherited checked exception that can be coupled with more underneath functions.